### PR TITLE
Start leveraging MachineInfo with for_* methods

### PR DESCRIPTION
**Pull request by Ericson2314**
Trying to break off pieces of #4010.

### CODE REVIEW

**Comment by jpakkane** [comment]
The `NotImplemented` seems a bit suspicious. It leads to things like this:

```python
x = ... # machineinfo object
y = ... # an object of some other class
if x == y:
  print('This gets printed')
```

The reason it gets printed is that the comparison returns a `NotImplemented` class object and it evaluates to `true` in Python.

**Comment by Ericson2314** [comment]
I believe per https://docs.python.org/3/library/constants.html#NotImplemented that that is not the case.
```python
x = ... # machineinfo object
y = ... # an object of some other class
if x.__eq__(y):
  print('This gets printed')
if y.__eq__(x):
  print('This gets printed')
if x == y:
  print('This does *not* get printed')
```

c.f.
```python
>>> (0).__eq__(None)
NotImplemented
>>> None.__eq__(0)
NotImplemented
>>> 0 == None
False
```

File: mesonbuild/mesonlib.py [hunk 1]
```diff
@@ -277,6 +279,53 @@ def classify_unity_sources(compilers, sources):
...
+
+MachineChoice = OrderedEnum('MachineChoice', ['BUILD', 'HOST', 'TARGET'])
+
+class PerMachine:
+    def __init__(self, build, host, target):
+        self.build = build,
+        self.host = host
+        self.target = target
+
+    def __getitem__(self, machine: MachineChoice):
```
Comment (jpakkane) [inline mesonbuild/mesonlib.py]: We still have to support plain Python 3.5. That means no type annotations. Unfortunately.

Comment (Ericson2314) [reply mesonbuild/mesonlib.py]: It doesn't enforce anything in 3.5 but the syntax (including name resolution) is supported. Try:
```python
>>> def a(b: int):
...     return b
...
>>> a('a')
'a'
>>> def a(b: iint):
...     return b
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'iint' is not defined
```
I find even without enforcement it is still useful for documentation's sake.

**Comment by Ericson2314** [comment]
Thanks!

**Pull request merged**